## Importing the necessary python libraries for the quiz

In [1]:
import pandas as pd

import numpy as np

## Importing the dataset for the Quiz

In [2]:
fuel_dataset = pd.read_csv("hamoye_dataset.csv", encoding='ISO-8859-1')

# Question 1:
### Create a list by combining the elements of A = [1,2,3,4,5,6] and B = [13,21,34]

In [3]:
A = [1,2,3,4,5,6]

B = [13,21,34]

In [4]:
A.extend(B)

print(A)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


# Question 2:

### Creating an identity element

In [5]:
iden_matrix = np.identity(3)

iden_matrix

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# Question 3:
### Which fuel type code has the lowest average fuel cost per unit burned?

In [6]:
fuel_dataset.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean().sort_values()

# Grouping the dataset by 'fuel_type_code_pud1' considering 'fuel_cost_per_unit_burned'

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

### From the above table, the fuel type code whith the lowest average fuel cost per unit burned is gas.



# Question 4:
### What is the standard deviation and 75th percentile of the measure of energy per unit in 2 decimal places?

In [7]:
fuel_dataset['fuel_mmbtu_per_unit'].describe()

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

In [8]:
std = fuel_dataset['fuel_mmbtu_per_unit'].describe()['std']

print(f"The standard deviation of the measure of the energy per unit, in two decimal places = {std:.2f}")

The standard deviation of the measure of the energy per unit, in two decimal places = 10.60


In [9]:
percentile = fuel_dataset['fuel_mmbtu_per_unit'].describe()['75%']

print(f"The 75th percentile of the measure of the energy per unit, in two decimal places = {percentile:.2f}")

The 75th percentile of the measure of the energy per unit, in two decimal places = 17.01


# Question 5:
### What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [10]:
skewness = fuel_dataset['fuel_qty_burned'].skew()

print(f"The skewness for the fuel quantity burned in two decimal places = {skewness:.2f}")

The skewness for the fuel quantity burned in two decimal places = 15.85


In [11]:
kurtosis = fuel_dataset['fuel_qty_burned'].kurtosis()

print(f"The kurtosis for the fuel quantity burned in two decimal places = {kurtosis:.2f}")

The kurtosis for the fuel quantity burned in two decimal places = 651.37


# Question 6:
### Which feature has missing values? What is the total number of missing values? What is the percentage of the missing rows as a factor of the total number of rows in three decimal places?

In [12]:
fuel_dataset.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

From the above, fuel_unit is the feature with missing values.

There are 180 missing values in total.

In [13]:
total_rows = len(fuel_dataset)

In [14]:
percent_missing_rows = (180/total_rows)*100

In [15]:
print(f"The percentage of missing rows as a factor of the total number of rows = {percent_missing_rows:.3f}%")

The percentage of missing rows as a factor of the total number of rows = 0.610%


# Question 7:
### What category does the feature with missing values fall?
### What missing value imputatation technique would you use?

In [16]:
# Category of fuel_unit: Categorical
# Imputation technique: mode (mcf)

# Question 8:
### Which of the features have the second lowest and third loweast correlation with the fuel cost per unit burned?

In [17]:
fuel_dataset.corr().abs().sort_values(by ='fuel_cost_per_unit_burned')

utility_id_ferc1  report_year  fuel_qty_burned  \
fuel_cost_per_mmbtu                   0.006122     0.010261         0.001896   
fuel_mmbtu_per_unit                   0.066946     0.110853         0.080946   
fuel_cost_per_unit_delivered          0.016414     0.014043         0.003551   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                       0.057447     0.012952         1.000000   
utility_id_ferc1                      1.000000     0.093323         0.057447   
fuel_cost_per_unit_burned             0.037863     0.013599         0.018535   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
fuel_cost_per_mmbtu                      0.005884                   0.000437   
fuel_mmbtu_per_unit                      1.000000                   0.010034   
fuel_cost_per_unit_delivered             0.009039                   0.011007   
report_year                              0.110853                   0.013599   
fuel_qty_burned                          0.080946                   0.018535   
utility_id_ferc1                         0.066946                   0.037863   
fuel_cost_per_unit_burned                0.010034                   1.000000   

                              fuel_cost_per_unit_delivered  \
fuel_cost_per_mmbtu                               0.000109   
fuel_mmbtu_per_unit                               0.009039   
fuel_cost_per_unit_delivered                      1.000000   
report_year                                       0.014043   
fuel_qty_burned                                   0.003551   
utility_id_ferc1                                  0.016414   
fuel_cost_per_unit_burned                         0.011007   

                              fuel_cost_per_mmbtu  
fuel_cost_per_mmbtu                      1.000000  
fuel_mmbtu_per_unit                      0.005884  
fuel_cost_per_unit_delivered             0.000109  
report_year                              0.010261  
fuel_qty_burned                          0.001896  
utility_id_ferc1                         0.006122  
fuel_cost_per_unit_burned                0.000437

From the above table, the second and third lowest correlation under the fuel_cost_per_unit_burned column are 0.010034 and 0.011007 respectively, corresponding to fuel_mmbtu_per_unit and fuel_cost_per_unit_delivered.

# Question 9:
### For the fuel type coal, what is the percentage change in the fuel cost burned in 1998 compared to 1994?

In [18]:
fuel_type_grouped = fuel_dataset.groupby(['fuel_type_code_pudl', 'report_year'])['fuel_cost_per_unit_burned'].sum()

fuel_type_grouped

fuel_type_code_pudl  report_year
coal                 1994           14984.572
                     1995           14571.785
                     1996           12694.803
                     1997           43742.178
                     1998           11902.597
                                      ...    
waste                2014             591.594
                     2015             629.270
                     2016             539.079
                     2017             452.079
                     2018             660.972
Name: fuel_cost_per_unit_burned, Length: 148, dtype: float64

In [19]:
unit_cost_1994 = fuel_type_grouped['coal', 1994]

unit_cost_1998 = fuel_type_grouped['coal', 1998]

In [20]:
percent_change = ((unit_cost_1998 - unit_cost_1994)/unit_cost_1994)*100

print(f"The fuel cost per unit burned in 1998 is {abs(percent_change):.2f}% less than that in 1994")

The fuel cost per unit burned in 1998 is 20.57% less than that in 1994


# Question 10:
### Which year has the highest average fuel cost per unit delivered?

In [22]:
fuel_dataset.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

From the table above, the highest average fuel cost per unit delivered is 11140.197239, corresponding to the year 1997

# THANKS!